initial hpx threading support for libcds data structure

weilewei opened the pull request:
This PR 
- might fixe #2235
- is part of work contributing to [Concurrent Data Structure project](https://github.com/STEllAR-GROUP/hpx/wiki/Google-Summer-of-Code-%28GSoC%29-2020#concurrent-data-structure-support) funded through [Google Summer of Code 2020](https://summerofcode.withgoogle.com/projects/#6740781068451840)

## Proposed Changes

  - Add additional hpx thread data structure for libcds thread data structure guarded by `#ifdef HPX_HAVE_LIBCDS`. The support is particularly targeting Hazard Pointer, and Dynamic HP. RCU is still work in progress, however, it does not impact HP integration and current PR. 
  - Enable cmake content fetch to libcds as a subdirectory
  - Add overhead test for libcds HP, which is located in tests/performance/local

## Any background context you want to provide?
To build libcds in HPX, one might need to follow these steps
```
cmake -DHPX_WITH_LIBCDS=ON /path/to/hpx
```
To run the overhead test, one might do

```
make tests.performance.local.libcds_hazard_pointer_overhead
cd bin
./tests.performance.local.libcds_hazard_pointer_overhead
```

msimberg commented:
add to whitelist

msimberg commented on tests/performance/local/CMakeLists.txt:
```diff
@@ -25,6 +25,11 @@ if(NOT HPX_WITH_SANITIZERS)
 
+if(HPX_WITH_LIBCDS)
+  set(benchmarks ${benchmarks} libcds_hp_overhead)
+  set(libcds_hp_overhead_FLAGS DEPENDENCIES iostreams_component hpx_timing)
```
```suggestion
  set(libcds_hp_overhead_FLAGS DEPENDENCIES iostreams_component)
```

Linking to individual modules like that should only be done if `NOLIBS` is used.

msimberg commented on tests/performance/local/CMakeLists.txt:
```diff
@@ -216,6 +221,11 @@ foreach(benchmark ${benchmarks})
 endforeach()
 
+if(HPX_WITH_LIBCDS)
+  target_compile_definitions(cds PUBLIC CDS_THREADING_HPX)
```
Should this be set here? Why not in `HPX_SetupLibCDS` or even in CDS itself?

msimberg commented on tests/performance/local/CMakeLists.txt:
```diff
@@ -216,6 +221,11 @@ foreach(benchmark ${benchmarks})
 
+if(HPX_WITH_LIBCDS)
+  target_compile_definitions(cds PUBLIC CDS_THREADING_HPX)
+  target_link_libraries(libcds_hp_overhead_test PRIVATE cds hpx_core)
```
```suggestion
  target_link_libraries(libcds_hp_overhead_test PRIVATE cds)
```

No need for `hpx_core`, it should get linked to HPX by `add_hpx_executable` above.

msimberg commented on tests/performance/local/libcds_hp_overhead.cpp:
```diff
@@ -0,0 +1,345 @@
+//  http://www.boost.org/LICENSE_1_0.txt)
+
+#include <cds/gc/hp.h>    // for cds::HP (Hazard Pointer) SMR
+#include <cds/init.h>     // for cds::Initialize and cds::Terminate
```
We usually put external headers after HPX headers and before standard library headers.

msimberg commented on tests/performance/local/libcds_hp_overhead.cpp:
```diff
@@ -0,0 +1,345 @@
+{
+#ifdef CDS_THREADING_HPX
+    if (uselibcds)
+        cds::threading::Manager::attachThread();
```
Could this not be added to `future_overhead.cpp` instead?

weilewei replied:
If we move libcds test to future_overhead.cpp, then this file might need `#ifdef HPX_HAVE_LIBCDS` and/or `#ifdef CDS_THREADING_HPX` everywhere, and also need add an extra bool input argument `bool libcds` to some measuring functions as well. I wonder will these actions become messy to future_overhead test? Just some thoughts.

msimberg commented on libs/coroutines/include/hpx/coroutines/coroutine.hpp:
```diff
@@ -96,6 +97,19 @@ namespace hpx { namespace threads { namespace coroutines {
         }
 
+#ifdef HPX_HAVE_LIBCDS
+        std::array<std::size_t, 3> get_libcds_data() const
```
A typedef for `std::array<std::size_t, 3>` might be useful.

weilewei replied:
Sure, I will fix this later today.

weilewei replied:
Ok, now it is fixed.

hkaiser commented on CMakeLists.txt:
```diff
@@ -930,6 +930,14 @@ if(HPX_WITH_APEX)
+if(HPX_WITH_LIBCDS)
+  hpx_add_config_define(HPX_HAVE_LIBCDS) # tell HPX that we use LibCDS
+  include(HPX_SetupLibCDS)
+endif()
```
Is there some error handling?

weilewei replied:
Added

hkaiser commented on libs/coroutines/include/hpx/coroutines/coroutine.hpp:
```diff
@@ -96,6 +97,19 @@ namespace hpx { namespace threads { namespace coroutines {
+            return impl_.get_libcds_data();
+        }
+
+        cds_type set_libcds_data(cds_type data)
```
You might want to pass the argument as `const&` to avoid unnecessary copies.

hkaiser commented on libs/coroutines/include/hpx/coroutines/detail/context_base.hpp:
```diff
@@ -85,6 +88,12 @@ namespace hpx { namespace threads { namespace coroutines { namespace detail {
+          , libcds_thread_data(cds_type{
+                { 0,
+                    0,
+                    0 }})
```
This formatting seems to be off.

hkaiser commented on libs/coroutines/include/hpx/coroutines/detail/context_base.hpp:
```diff
@@ -98,6 +107,9 @@ namespace hpx { namespace threads { namespace coroutines { namespace detail {
             m_thread_data = 0;
+#endif
+#ifdef HPX_HAVE_LIBCDS
+            libcds_thread_data = cds_type{{ 0, 0, 0 }};
```
Formatting off?

hkaiser commented on libs/coroutines/include/hpx/coroutines/detail/context_base.hpp:
```diff
@@ -229,6 +244,20 @@ namespace hpx { namespace threads { namespace coroutines { namespace detail {
+            return libcds_thread_data;
+        }
+
+        cds_type set_libcds_data(cds_type data)
```
Same as above, pass by `const&`.

hkaiser commented on libs/coroutines/include/hpx/coroutines/coroutine.hpp:
```diff
@@ -96,6 +97,19 @@ namespace hpx { namespace threads { namespace coroutines {
 
+#ifdef HPX_HAVE_LIBCDS
+        typedef std::array<std::size_t, 3> cds_type;
+        cds_type get_libcds_data() const
```
Return by `const&` to avoid copies?

hkaiser commented on libs/coroutines/include/hpx/coroutines/stackless_coroutine.hpp:
```diff
@@ -121,6 +122,21 @@ namespace hpx { namespace threads { namespace coroutines {
         }
 
+#ifdef HPX_HAVE_LIBCDS
+        typedef std::array<std::size_t, 3> cds_type;
```
Should all uses of `cds_type` refer t the same using declaration?

hkaiser commented on libs/coroutines/include/hpx/coroutines/stackless_coroutine.hpp:
```diff
@@ -121,6 +122,21 @@ namespace hpx { namespace threads { namespace coroutines {
+            libcds_data_ = data;
+            return olddata;
+        }
+#endif
```
Could these functions go into the base class instead (avoids virtual functions and code duplication)?

weilewei replied:
Would you mind telling which file specifically?

hkaiser replied:
Please ignore this comment, I was mistaken.

weilewei commented on libs/threading_base/src/thread_helpers.cpp:
```diff
@@ -204,6 +205,33 @@ namespace hpx { namespace threads {
+        {
+            HPX_THROWS_IF(ec, null_thread_id, "hpx::threads::get_libcds_data",
+                "null thread id encountered");
+            return cds_type{{0, 0, 0}};
```
@hkaiser I agreed with passing by reference and return by reference, but what should I do for this return which returns a copy of cds data if `HPX_UNLIKELY(!id) == true`? Any suggestion?

hkaiser replied:
What about:
```
static cds_type data({{0, 0, 0}});
return data;
```
?

weilewei commented:
I modified the underlying data structure for thread data for libcds, removed std::array<std::size_t, 3>, and now use three individual elements representing thread different thread-private data: thread manager, hazard pointer, dynamic hazard pointers. 

Because these three elements are independent of each other, when writing and reading one of the elements, we really do not need to write/read three elements together.

hkaiser commented on libs/coroutines/include/hpx/coroutines/coroutine.hpp:
```diff
@@ -98,16 +97,35 @@ namespace hpx { namespace threads { namespace coroutines {
         }
 
-        cds_type set_libcds_data(cds_type data)
+        std::size_t& set_libcds_data(std::size_t& data)
```
What's your rationale of using references for the parameter and return value in this case?

weilewei replied:
Oh, should be a typo, I meant const& everywhere, but I forgot to add it, will add soon.

hkaiser replied:
That makes it only slightly better. Why do you think you need references in the first place?

weilewei replied:
Oh, because you commented here saying adding const references to avoid copy: https://github.com/STEllAR-GROUP/hpx/pull/4823#discussion_r453971287 and https://github.com/STEllAR-GROUP/hpx/pull/4823#discussion_r453972807

hkaiser replied:
@weilewei `sizeof(std::array<3, size_t>) == 3 * sizeof(std::array<3, size_t> const&)`, however `sizeof(size_t) == sizeof(size_t const&)`. Thus, if there is no size benefit of using references, then value semantics are preferable any day. Moreover, returning a reference from a function forces you to use a function-local static in the implementation which will create a data race as this variable would be shared in between threads.

weilewei replied:
Thanks for the clarification, ok, then I will stick to value semantics then, pass by value and return by value.

hkaiser commented:
> I modified the underlying data structure for thread data for libcds, removed std::array<std::size_t, 3>, and now use three individual elements representing thread different thread-private data: thread manager, hazard pointer, dynamic hazard pointers.
> 
> Because these three elements are independent of each other, when writing and reading one of the elements, we really do not need to write/read three elements together. This is much easier for retrieving data, debugging, and providing straight-forward API for libcds end.

I like this change. Could we use more descriptive function names, however? `libcds_data`, `libcds_hazard_pointer_data`, and `libcds_dynamic_hazard_pointer_data` sounds good enough for me, but anything else descriptive would be fine as well.

weilewei commented:
Please review it again, thanks

hkaiser commented on libs/coroutines/include/hpx/coroutines/detail/context_base.hpp:
```diff
@@ -48,6 +48,7 @@
 #include <hpx/coroutines/detail/tss.hpp>
 #include <hpx/coroutines/thread_id_type.hpp>
 
+#include <array>
```
Do we still need this `#include <array>`?

hkaiser commented on libs/coroutines/include/hpx/coroutines/detail/context_base.hpp:
```diff
@@ -204,6 +214,11 @@ namespace hpx { namespace threads { namespace coroutines { namespace detail {
+            libcds_data_ = 0;
+            libcds_hazard_pointer_data_ = 0;
+            libcds_dynamic_hazard_pointer_data_ = 0;
+#endif
```
I think this #if/#endif should be outside the `HPX_HAVE_THREAD_LOCAL_STORAGE` condition above.

hkaiser commented on libs/coroutines/include/hpx/coroutines/detail/context_base.hpp:
```diff
@@ -229,6 +244,44 @@ namespace hpx { namespace threads { namespace coroutines { namespace detail {
+
+        std::size_t set_libcds_data(std::size_t data)
+        {
+            static std::size_t olddata = libcds_data_;
```
This creates a data race as it is shared amongst threads. Same is true for the other functions below.

hkaiser replied:
If in doubt, you could write it as:
```
std::size_t set_libcds_data(std::size_t data)
{
    std::swap(data, libcds_data_);
    return data;
}
```

hkaiser commented on libs/coroutines/include/hpx/coroutines/detail/context_base.hpp:
```diff
@@ -345,6 +403,11 @@ namespace hpx { namespace threads { namespace coroutines { namespace detail {
+        mutable std::size_t libcds_data_;
+        mutable std::size_t libcds_hazard_pointer_data_;
+        mutable std::size_t libcds_dynamic_hazard_pointer_data_;
+#endif
```
This should be outside the condition `HPX_HAVE_THREAD_LOCAL_STORAGE` above.

hkaiser commented on libs/coroutines/include/hpx/coroutines/stackless_coroutine.hpp:
```diff
@@ -18,6 +18,7 @@
 #include <hpx/functional/unique_function.hpp>
 #include <hpx/type_support/unused.hpp>
 
+#include <array>
```
Do we still need this `#include <array>`?

hkaiser commented on libs/coroutines/include/hpx/coroutines/stackless_coroutine.hpp:
```diff
@@ -121,6 +122,44 @@ namespace hpx { namespace threads { namespace coroutines {
+
+        std::size_t set_libcds_data(std::size_t data)
+        {
+            static std::size_t olddata = libcds_data_;
```
Same problem as above: the function-local static will create a data race.

hkaiser commented on libs/threading_base/include/hpx/threading_base/thread_helpers.hpp:
```diff
@@ -20,6 +20,7 @@
 #include <hpx/threading_base/thread_pool_base.hpp>
 #include <hpx/timing/steady_clock.hpp>
 
+#include <array>
```
Still needed?

hkaiser commented on tests/performance/local/libcds_hp_overhead.cpp:
```diff
@@ -0,0 +1,344 @@
+///////////////////////////////////////////////////////////////////////////////
+double null_function(bool uselibcds) noexcept
+{
+#ifdef CDS_THREADING_HPX
```
If libCDS is not configured to support HPX threads this test should not be enabled in the first place.

msimberg commented on libs/coroutines/include/hpx/coroutines/detail/context_base.hpp:
```diff
@@ -204,6 +214,11 @@ namespace hpx { namespace threads { namespace coroutines { namespace detail {
 #else
             m_thread_data = 0;
+#endif
+#ifdef HPX_HAVE_LIBCDS
```
Minor style detail: We tend to use `#if defined(X)` instead of `#ifdef X`.

weilewei replied:
Thanks, I just adapted the style. Please check again.

msimberg commented:
retest

msimberg commented:
retest
